tests/integration/ha_tests/test_functional.py

```diff
@@ -0,0 +1,289 @@
+            "access-key": github_secrets["AWS_ACCESS_KEY"],
+            "secret-key": github_secrets["AWS_SECRET_KEY"],
+        },
+    }
+    yield configs, credentials
+    # Delete the previously created objects.
+    logger.info("deleting the previously created backups")
+    for cloud, config in configs.items():
+        session = boto3.session.Session(
+            aws_access_key_id=credentials[cloud]["access-key"],
+            aws_secret_access_key=credentials[cloud]["secret-key"],
+            region_name=config["region"],
+        )
+        s3 = session.resource(
+            "s3", endpoint_url=construct_endpoint(config["endpoint"], config["region"])
+        )
+        bucket = s3.Bucket(config["bucket"])
+        for bucket_object in bucket.objects.filter(Prefix=config["path"].lstrip("/")):
+            bucket_object.delete()
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_instance_graceful_restart(ops_test: OpsTest, charm: str) -> None:
+    """Test graceful restart of a service."""
+    async with ops_test.fast_forward():
+        # Deploy the charm.
+        logger.info("deploying charm")
+        await ops_test.model.deploy(
+            charm,
+            application_name=APPLICATION_NAME,
+            num_units=1,
+            series=CHARM_SERIES,
+            config={"profile": "testing"},
+        )
+
+        logger.info("waiting for idle")
+        await ops_test.model.wait_for_idle(apps=[APPLICATION_NAME], status="active", timeout=1500)
+        assert ops_test.model.applications[APPLICATION_NAME].units[0].workload_status == "active"
+
+        primary_name = ops_test.model.applications[APPLICATION_NAME].units[0].name
+
+        logger.info("waiting for postgresql")
+        for attempt in Retrying(stop=stop_after_delay(15 * 3), wait=wait_fixed(3), reraise=True):
+            with attempt:
+                assert await is_postgresql_ready(ops_test, primary_name)
+
+        # Get unit hostname and IP.
+        logger.info("wait for hostname")
+        primary_hostname = await get_machine_from_unit(ops_test, primary_name)
+
+        logger.info("restarting service")
+        await lxc_restart_service(primary_hostname)
+
+        logger.info("waiting for idle")
+        await ops_test.model.wait_for_idle(
+            apps=[APPLICATION_NAME], status="active", timeout=1500, raise_on_error=False
+        )
+        assert ops_test.model.applications[APPLICATION_NAME].units[0].workload_status == "active"
+
+        logger.info("check graceful shutdown")
+        assert await check_graceful_shutdown(ops_test, primary_name)
+
+        logger.info("check success recovery")
+        assert await check_success_recovery(ops_test, primary_name)
+
+        logger.info("remove application")
+        for attempt in Retrying(stop=stop_after_delay(15 * 3), wait=wait_fixed(3), reraise=True):
+            with attempt:
+                await ops_test.model.remove_application(APPLICATION_NAME, block_until_done=True)
```

Comment:
Testing on Juju 3.5.1, I often faced the following error:
```sh
unit-postgresql-test-app-0: 14:50:09 ERROR juju.worker.uniter.operation hook "pgdata-storage-detaching" (via hook dispatching script: dispatch) failed: exit status 1
unit-postgresql-test-app-0: 14:55:11 ERROR unit.postgresql-test-app/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/./src/charm.py", line 1645, in <module>
    main(PostgresqlOperatorCharm)
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/venv/ops/main.py", line 544, in main
    manager.run()
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/venv/ops/main.py", line 520, in run
    self._emit()
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/venv/ops/main.py", line 509, in _emit
    _emit_charm_event(self.charm, self.dispatcher.event_name)
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/venv/ops/main.py", line 143, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/venv/ops/framework.py", line 851, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/venv/ops/framework.py", line 941, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/./src/charm.py", line 432, in _on_pgdata_storage_detaching
    primary = self._patroni.get_primary(unit_name_pattern=True)
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/./src/charm.py", line 705, in _patroni
    self._peer_members_ips,
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/./src/charm.py", line 735, in _peer_members_ips
    addresses = self.members_ips
  File "/var/lib/juju/agents/unit-postgresql-test-app-0/charm/./src/charm.py", line 756, in members_ips
    return set(json.loads(self._peers.data[self.app].get("members_ips", "[]")))
AttributeError: 'NoneType' object has no attribute 'data'
The test got stuck at this point.
```

Reply:
Sure.
* Juju: 3.4.4
* LXD: 5.21.1-d46c406
* Charmcraft: 2.7.0
* Snap: 2.63+24.04

Reply:
Could you test on Juju 3.5.1? Like it was in original issue by you? 
I want to know version of LXD, Charmcraft, and Snap on witch this error is reproducible.

Reply:
Sure. I tested again with Juju 3.5.1. Those are the version I used:
* Juju: 3.5.1
* LXD: 5.21.1-d46c406
* Charmcraft: 2.7.0
* Snap: 2.63+24.04

I also copied this branch to another PR to test it. The error is the following: https://github.com/canonical/postgresql-operator/actions/runs/9881101326/job/27384074248#step:28:2995